### PR TITLE
fix(ui): Prevent error boundary setting state after unmount

### PR DESCRIPTION
**Pull request by scttcper**
Sometimes during development the error boundary component errors when changing pages.

The browserHistory listener is still triggered sometimes after `unlistenBrowserHistory` is called and tries to set state to null even though it is being destroyed.

image

### CODE REVIEW

File: src/sentry/static/sentry/app/components/errorBoundary.tsx [hunk 1]
```diff
@@ -70,13 +77,14 @@ class ErrorBoundary extends React.Component<Props, State> {
...
+    this._isMounted = false;
     if (this.unlistenBrowserHistory) {
       this.unlistenBrowserHistory();
     }
   }
 
-  // XXX: browserHistory.listen does not have a correct return type.
-  unlistenBrowserHistory: any;
+  unlistenBrowserHistory: ReturnType<typeof browserHistory.listen> | undefined;
+  _isMounted: boolean | undefined;
```
Comment (AbhiPrasad) [inline src/sentry/static/sentry/app/components/errorBoundary.tsx]: Why not just type this as a Boolean and instantiate as false?

**Review by AbhiPrasad — APPROVED** [review]
Looks right to me.

File: src/sentry/static/sentry/app/components/errorBoundary.tsx (hunk 1 above)
Comment (scttcper) [reply src/sentry/static/sentry/app/components/errorBoundary.tsx]: thanks, simplified the types here a bit

File: src/sentry/static/sentry/app/components/errorBoundary.tsx [hunk 2]
```diff
@@ -70,13 +75,14 @@ class ErrorBoundary extends React.Component<Props, State> {
...
+    this._isMounted = false;
     if (this.unlistenBrowserHistory) {
       this.unlistenBrowserHistory();
     }
   }
 
-  // XXX: browserHistory.listen does not have a correct return type.
-  unlistenBrowserHistory: any;
+  private unlistenBrowserHistory?: ReturnType<typeof browserHistory.listen>;
+  private _isMounted = false;
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/components/errorBoundary.tsx:85]: We pretty rarely us the visibility keywords

Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/components/errorBoundary.tsx:85]: We pretty rarely use visibility keywords

File: src/sentry/static/sentry/app/components/errorBoundary.tsx [hunk 3]
```diff
@@ -70,13 +75,14 @@ class ErrorBoundary extends React.Component<Props, State> {
...
   componentWillUnmount() {
+    this._isMounted = false;
     if (this.unlistenBrowserHistory) {
       this.unlistenBrowserHistory();
     }
   }
 
-  // XXX: browserHistory.listen does not have a correct return type.
-  unlistenBrowserHistory: any;
+  private unlistenBrowserHistory?: ReturnType<typeof browserHistory.listen>;
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/components/errorBoundary.tsx:84]: Thank you for typing this 😍

**Review by billyvg — APPROVED** [review]

**Comment by billyvg** [comment]
> boundry

File: src/sentry/static/sentry/app/components/errorBoundary.tsx [hunk 4]
```diff
@@ -70,13 +75,14 @@ class ErrorBoundary extends React.Component<Props, State> {
...
+    this._isMounted = false;
     if (this.unlistenBrowserHistory) {
       this.unlistenBrowserHistory();
     }
   }
 
-  // XXX: browserHistory.listen does not have a correct return type.
-  unlistenBrowserHistory: any;
+  unlistenBrowserHistory?: ReturnType<typeof browserHistory.listen>;
+  _isMounted = false;
```
Comment (leedongwei) [inline src/sentry/static/sentry/app/components/errorBoundary.tsx:85]: Would it make sense to set these as private properties.

File: src/sentry/static/sentry/app/components/errorBoundary.tsx (hunk 3 above)
Comment (leedongwei) [reply src/sentry/static/sentry/app/components/errorBoundary.tsx:84]: I use it on oddball properties/methods that should really stay inside a component. 

IIRC it doesn't stop someone from passing it to a child component but it's an additional hint that they shouldn't do it.

File: src/sentry/static/sentry/app/components/errorBoundary.tsx (hunk 2 above)
Comment (leedongwei) [reply src/sentry/static/sentry/app/components/errorBoundary.tsx:85]: I use it on oddball properties/methods that should really stay inside a component.

IIRC it doesn't stop someone from passing it to a child component but it's an additional hint that they shouldn't do it.

**Review by leedongwei — APPROVED** [review]
LGTM. I'm not strongly opinionated on setting that property/method to `private`.

**Review by matejminar — APPROVED** [review]

**Pull request merged**